// @rpldy/uploader/lib/batch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _shared = require("@rpldy/shared");

var _defaults = require("./defaults");

var bCounter = 0;

var processFiles = function processFiles(batchId, files, fileFilter) {
  return Array.prototype.filter.call(files, fileFilter || _defaults.DEFAULT_FILTER).map(function (f) {
    return (0, _shared.createBatchItem)(f, batchId);
  });
};

var _default = function _default(files, uploaderId, options) {
  bCounter += 1;
  var id = "batch-".concat(bCounter); //in case files list was created in a different context(window) need to check toString

  var isFileList = files instanceof FileList || files.toString() === "[object FileList]";
  files = Array.isArray(files) || isFileList ? files : [files];
  return {
    id: id,
    uploaderId: uploaderId,
    items: processFiles(id, files, options.fileFilter),
    state: _shared.BATCH_STATES.ADDED,
    completed: 0,
    loaded: 0
  };
};

exports["default"] = _default;
